<a href="https://colab.research.google.com/github/hdomin/Python/blob/main/06_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ejercicio práctico

[Link al Curso Udemy](https://www.udemy.com/course/data-science-bootcamp-with-python/learn/quiz/5211368#learning-tools)

In [ ]:
import pandas as pd


El archivo fifa.csv lo puede encontrar [aquí](https://github.com/hdomin/Python/raw/main/fifa.csv)

In [ ]:
df = pd.read_csv('sample_data/fifa.csv')
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club  ... Composure Marking StandingTackle  SlidingTackle  \
0         FC Barcelona  ...      96.0    33.0           28.0           26.0   
1             Juventus  ...      95.0    28.0           31.0           23.0   
2  Paris Saint-Germain  ...      94.0    27.0           24.0           33.0   
3    Manchester United  ...      68.0    15.0           21.0           13.0   
4      Manchester City  ...      88.0    68.0           58.0           51.0   

  GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes Release Clause  
0      6.0        11.0       15.0           14.0        8.0        €226.5M  
1      7.0        11.0       15.0           14.0       11.0        €127.1M  
2      9.0         9.0       15.0           15.0       11.0        €228.1M  
3     90.0        85.0       87.0           88.0       94.0        €138.6M  
4     15.0        13.0        5.0           10.0       13.0        €196.4M  

[5 rows x 89 columns]

Revisa todas las columnas del data set

In [ ]:
df.columns

Index(['Unnamed: 0', 'ID', 'Name', 'Age', 'Photo', 'Nationality', 'Flag',
       'Overall', 'Potential', 'Club', 'Club Logo', 'Value', 'Wage', 'Special',
       'Preferred Foot', 'International Reputation', 'Weak Foot',
       'Skill Moves', 'Work Rate', 'Body Type', 'Real Face', 'Position',
       'Jersey Number', 'Joined', 'Loaned From', 'Contract Valid Until',
       'Height', 'Weight', 'LS', 'ST', 'RS', 'LW', 'LF', 'CF', 'RF', 'RW',
       'LAM', 'CAM', 'RAM', 'LM', 'LCM', 'CM', 'RCM', 'RM', 'LWB', 'LDM',
       'CDM', 'RDM', 'RWB', 'LB', 'LCB', 'CB', 'RCB', 'RB', 'Crossing',
       'Finishing', 'HeadingAccuracy', 'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingT

Se busca en cada fila los valores sin datos en columnas (axis=1), se cuentan y se presentan en orden descendente

Por ejemplo la fila con el índice 13260 tiene 75 columnas sin valor y la fila con índice 6661 tiene una columna sin valor.

In [ ]:
df.isnull().sum(axis=1).sort_values(ascending=False)

13260    75
13249    75
13276    75
13277    75
13278    75
         ..
6661      1
6662      1
6663      1
6664      1
18206     1
Length: 18207, dtype: int64

La siguiente forma, revisa si hay algún valor vacío o perdido "missing value" en una fila y devuelve el índice, similar al anterior, pero solamente devuelve el valor True y no cuenta las columnas. 

Como se obervará, la longitud del arreglo devuelto es de 18,207 registros, el mismo valor de la consulta anterior.

In [ ]:
df.isnull().any(axis=1)

0        True
1        True
2        True
3        True
4        True
         ... 
18202    True
18203    True
18204    True
18205    True
18206    True
Length: 18207, dtype: bool

La siguiente validación, verifica si exíste algún registro donde todas las columnas tengan valores vacíos, y como se puede verificar, no hay registros con todas las columnas vacías.

In [ ]:
df.isnull().all(axis=1).sum()

0

Analizando un poco los datos, podríamos obtener todos los registros que tienen más de 50 columnas sin datos

In [ ]:
df[ df.isnull().sum(axis=1)>50]

Unnamed: 0      ID            Name  Age  \
13236       13236  177971      J. McNulty   33   
13237       13237  195380      J. Barrera   29   
13238       13238  139317        J. Stead   35   
13239       13239  240437     A. Semprini   20   
13240       13240  209462      R. Bingham   24   
13241       13241  219702    K. Dankowski   21   
13242       13242  225590       I. Colman   23   
13243       13243  233782       M. Feeney   19   
13244       13244  239158        R. Minor   30   
13245       13245  242998          Klauss   21   
13246       13246  244022      I. Sissoko   22   
13247       13247  189238         F. Hart   28   
13248       13248  211511   L. McCullough   24   
13249       13249  224055       Li Yunqiu   27   
13250       13250  244535       F. Garcia   29   
13251       13251  134968    R. Haemhouts   34   
13252       13252  225336       E. Binaku   22   
13253       13253  171320       G. Miller   31   
13254       13254  246328      A. Aidonis   17   
13255       13255  196921        L. Sowah   25   
13256       13256  202809       R. Deacon   26   
13257       13257  226617   Jang Hyun Soo   25   
13258       13258  230713     A. Al Malki   23   
13259       13259  234809     E. Guerrero   27   
13260       13260  246073         Hernáiz   20   
13261       13261  221498   H. Al Mansour   25   
13262       13262  244026         H. Paul   24   
13263       13263  244538        S. Bauer   25   
13264       13264  201019      M. Chergui   29   
13265       13265  221499      D. Gardner   28   
13266       13266  237371    L. Bengtsson   20   
13267       13267  242491    F. Jaramillo   22   
13268       13268  153148      L. Garguła   37   
13269       13269  244540       S. Rivera   26   
13270       13270  245564        Vinicius   19   
13271       13271  213821    F. Sepúlveda   26   
13272       13272  240701       L. Spence   22   
13273       13273  242237      B. Lepistu   25   
13274       13274  244029     A. Abruscia   27   
13275       13275  244541     E. González   23   
13276       13276  211006      M. Al Amri   26   
13277       13277  215102    J. Rebolledo   26   
13278       13278  246078      C. Mamengi   17   
13279       13279  239679    P. Mazzocchi   22   
13280       13280  244543       Y. Ammour   19   
13281       13281  212800  Jwa Joon Hyeop   27   
13282       13282  231232      O. Marrufo   25   
13283       13283  232256     Han Pengfei   25   

                                                Photo       Nationality  \
13236  https://cdn.sofifa.org/players/4/19/177971.png          Scotland   
13237  https://cdn.sofifa.org/players/4/19/195380.png         Nicaragua   
13238  https://cdn.sofifa.org/players/4/19/139317.png           England   
13239  https://cdn.sofifa.org/players/4/19/240437.png             Italy   
13240  https://cdn.sofifa.org/players/4/19/209462.png           England   
13241  https://cdn.sofifa.org/players/4/19/219702.png            Poland   
13242  https://cdn.sofifa.org/players/4/19/225590.png         Argentina   
13243  https://cdn.sofifa.org/players/4/19/233782.png           England   
13244  https://cdn.sofifa.org/players/4/19/239158.png           Denmark   
13245  https://cdn.sofifa.org/players/4/19/242998.png            Brazil   
13246  https://cdn.sofifa.org/players/4/19/244022.png            France   
13247  https://cdn.sofifa.org/players/4/19/189238.png           Austria   
13248  https://cdn.sofifa.org/players/4/19/211511.png  Northern Ireland   
13249  https://cdn.sofifa.org/players/4/19/224055.png          China PR   
13250  https://cdn.sofifa.org/players/4/19/244535.png          Paraguay   
13251  https://cdn.sofifa.org/players/4/19/134968.png           Belgium   
13252  https://cdn.sofifa.org/players/4/19/225336.png           Albania   
13253  https://cdn.sofifa.org/players/4/19/171320.png          Scotland   
13254  https://cdn.sofifa.org/players/4/19/246328.png           Germany   
13255  https://cdn.sofifa.org/players/4/19/196921.png   

Antes de realizar cualquier sustitución o eliminación, verificamos el shape (dimensiones) y el tamaño de nuestro data frame, y obtenemos 18,207 registros con 89 columnas para un total de 1,620,423 elementos en nuestra matríz

In [ ]:
df.shape, df.size

((18207, 89), 1620423)

In [ ]:
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club  ... Composure Marking StandingTackle  SlidingTackle  \
0         FC Barcelona  ...      96.0    33.0           28.0           26.0   
1             Juventus  ...      95.0    28.0           31.0           23.0   
2  Paris Saint-Germain  ...      94.0    27.0           24.0           33.0   
3    Manchester United  ...      68.0    15.0           21.0           13.0   
4      Manchester City  ...      88.0    68.0           58.0           51.0   

  GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes Release Clause  
0      6.0        11.0       15.0           14.0        8.0        €226.5M  
1      7.0        11.0       15.0           14.0       11.0        €127.1M  
2      9.0         9.0       15.0           15.0       11.0        €228.1M  
3     90.0        85.0       87.0           88.0       94.0        €138.6M  
4     15.0        13.0        5.0           10.0       13.0        €196.4M  

[5 rows x 89 columns]

Ahora se realiza la validación de "missing values" por columna, muy similar a lo anterior, pero el dato resultante es validado por columna

In [ ]:
df.isnull().sum().sort_values(ascending=False)

Loaned From    16943
LWB             2085
LCM             2085
RS              2085
LW              2085
               ...  
Nationality        0
Photo              0
Age                0
Name               0
Unnamed: 0         0
Length: 89, dtype: int64

Si se desea observar la totalidad de las filas ejecutar la siguiente opción: 


In [ ]:
pd.set_option("max_rows", 89)
df.isnull().sum().sort_values(ascending=False)

Para analizar el porcentaje de "missing values" por cada columna, se genera un nuevo data frame y se pondera con el total de registros

In [ ]:
x = df.isnull().sum()
y = x / df.shape[0] * 100
z = { "Missing Values": x, "Percent" : y}
ndf = pd.DataFrame(z, columns=["Missing Values", "Percent"])
ndf.sort_values(by="Percent", ascending=False)

Missing Values    Percent
LDM                                 2085  11.451639
LS                                  2085  11.451639
RS                                  2085  11.451639
LW                                  2085  11.451639
LF                                  2085  11.451639
CF                                  2085  11.451639
RF                                  2085  11.451639
RW                                  2085  11.451639
LAM                                 2085  11.451639
CAM                                 2085  11.451639
RAM                                 2085  11.451639
LM                                  2085  11.451639
LCM                                 2085  11.451639
CM                                  2085  11.451639
RCM                                 2085  11.451639
RM                                  2085  11.451639
LWB                                 2085  11.451639
CDM                                 2085  11.451639
RDM                                 2085  11.451639
RWB                                 2085  11.451639
LB                                  2085  11.451639
LCB                                 2085  11.451639
CB                                  2085  11.451639
RCB                                 2085  11.451639
RB                                  2085  11.451639
ST                                  2085  11.451639
Release Clause                      1564   8.590103
Joined                              1553   8.529686
Contract Valid Until                 289   1.587302
Club                                 241   1.323667
Jersey Number                         60   0.329544
Position                              60   0.329544
ShotPower                             48   0.263635
Aggression                            48   0.263635
LongShots                             48   0.263635
Strength                              48   0.263635
Stamina                               48   0.263635
Jumping                               48   0.263635
Weight                                48   0.263635
Balance                               48   0.263635
Reactions                             48   0.263635
Interceptions                         48   0.263635
SprintSpeed                           48   0.263635
Acceleration                          48   0.263635
BallControl                           48   0.263635
Agility                               48   0.263635
Vision                                48   0.263635
Positioning                           48   0.263635
FKAccuracy                            48   0.263635
Penalties                             48   0.263635
Composure                             48   0.263635
Marking                               48   0.263635
StandingTackle                        48   0.263635
SlidingTackle                         48   0.263635
GKDiving                              48   0.263635
GKHandling                            48   0.263635
GKKicking                             48   0.263635
GKPositioning                         48   0.263635
GKReflexes                            48   0.263635
LongPassing                           48   0.263635
Volleys                               48   0.263635
Curve                                 48   0.263635
Height                                48   0.263635
ShortPassing                          48   0.263635
HeadingAccuracy                       48   0.263635
Finishing                             48   0.263635
Crossing                              48   0.263635
Real Face                             48   0.263635
Body Type                             48   0.263635
Work Rate                             48   0.263635
Skill Moves                           48   0.263635
Weak Foot                             48   0.263635
International Reputation              48   0.263635
Dribbling                             48   0.263635
Preferred Foot                        48   0.263635
Wage                                   0   0.000000
Special               

Al observar los datos, la columna "Loaned From" contiene un alto porcentaje de "missing values" por lo que sería recomendable eliminarla para el análisis si no es posible completarlos.  El comando lleva "axis=1" porque lo que se desea eliminar es la columna.

In [ ]:
df = df.drop(['Loaned From'], axis=1)

Y como se observa, ahora el data frame contiene 88 columnas, inicialmente 89.

In [ ]:
df.shape

(18207, 88)

Ahora se procederá a completar los valores vacíos del data frame

In [ ]:
df.dtypes[df.isnull().any()]

Preferred Foot               object
International Reputation    float64
Weak Foot                   float64
Skill Moves                 float64
Work Rate                    object
Body Type                    object
Real Face                    object
Height                       object
Weight                       object
LS                           object
ST                           object
RS                           object
LW                           object
LF                           object
CF                           object
RF                           object
RW                           object
LAM                          object
CAM                          object
RAM                          object
LM                           object
LCM                          object
CM                           object
RCM                          object
RM                           object
LWB                          object
LDM                          object
CDM                         

In [ ]:
df["Jersey Number"].fillna("NA", inplace=True)

Al validar nuevamente "Jersey Number", ya no tiene "missing values"

In [ ]:
df.dtypes[df.isnull().any()]

Otra forma de eliminar los "missing values" es con la muncion MODE

In [ ]:
df["Club"].mode()

0                   AS Monaco
1                     Arsenal
2             Atlético Madrid
3           Borussia Dortmund
4                     Burnley
5                  CD Leganés
6                Cardiff City
7                     Chelsea
8         Eintracht Frankfurt
9                      Empoli
10                    Everton
11               FC Barcelona
12         Fortuna Düsseldorf
13                  Frosinone
14                  Liverpool
15            Manchester City
16          Manchester United
17           Newcastle United
18                   RC Celta
19             Rayo Vallecano
20                Real Madrid
21                Southampton
22        TSG 1899 Hoffenheim
23          Tottenham Hotspur
24                Valencia CF
25    Wolverhampton Wanderers
dtype: object

In [ ]:
df["Club"] = df["Club"].fillna(df["Club"].mode()[0])

In [ ]:
df["Club"]

0               FC Barcelona
1                   Juventus
2        Paris Saint-Germain
3          Manchester United
4            Manchester City
                ...         
18202        Crewe Alexandra
18203         Trelleborgs FF
18204       Cambridge United
18205        Tranmere Rovers
18206        Tranmere Rovers
Name: Club, Length: 18207, dtype: object

In [ ]:
df["Position"] = df["Position"].fillna(df["Position"].mode()[0])
df["Joined"] = df["Joined"].fillna(df["Joined"].mode()[0])
df["Contract Valid Until"] = df["Contract Valid Until"].fillna(df["Contract Valid Until"].mode()[0])
df["Release Clause"] = df["Release Clause"].fillna(df["Release Clause"].mode()[0])

In [48]:
df['Preferred Foot'].fillna(0, inplace=True)
df['International Reputation'].fillna(0, inplace=True)
df['Weak Foot'].fillna(0, inplace=True)
df['Skill Moves'].fillna(0, inplace=True)
df['Work Rate'].fillna(0, inplace=True)
df['Body Type'].fillna(0, inplace=True)
df['Real Face'].fillna(0, inplace=True)
df['Height'].fillna(0, inplace=True)
df['Weight'].fillna(0, inplace=True)
df['LS'].fillna(0, inplace=True)
df['ST'].fillna(0, inplace=True)
df['RS'].fillna(0, inplace=True)
df['LW'].fillna(0, inplace=True)
df['LF'].fillna(0, inplace=True)
df['CF'].fillna(0, inplace=True)
df['RF'].fillna(0, inplace=True)
df['RW'].fillna(0, inplace=True)
df['LAM'].fillna(0, inplace=True)
df['CAM'].fillna(0, inplace=True)
df['RAM'].fillna(0, inplace=True)
df['LM'].fillna(0, inplace=True)
df['LCM'].fillna(0, inplace=True)
df['CM'].fillna(0, inplace=True)
df['RCM'].fillna(0, inplace=True)
df['RM'].fillna(0, inplace=True)
df['LWB'].fillna(0, inplace=True)
df['LDM'].fillna(0, inplace=True)
df['CDM'].fillna(0, inplace=True)
df['RDM'].fillna(0, inplace=True)
df['RWB'].fillna(0, inplace=True)
df['LB'].fillna(0, inplace=True)
df['LCB'].fillna(0, inplace=True)
df['CB'].fillna(0, inplace=True)
df['RCB'].fillna(0, inplace=True)
df['RB'].fillna(0, inplace=True)
df['Crossing'].fillna(0, inplace=True)
df['Finishing'].fillna(0, inplace=True)
df['HeadingAccuracy'].fillna(0, inplace=True)
df['ShortPassing'].fillna(0, inplace=True)
df['Volleys'].fillna(0, inplace=True)
df['Dribbling'].fillna(0, inplace=True)
df['Curve'].fillna(0, inplace=True)
df['FKAccuracy'].fillna(0, inplace=True)
df['LongPassing'].fillna(0, inplace=True)
df['BallControl'].fillna(0, inplace=True)
df['Acceleration'].fillna(0, inplace=True)
df['SprintSpeed'].fillna(0, inplace=True)
df['Agility'].fillna(0, inplace=True)
df['Reactions'].fillna(0, inplace=True)
df['Balance'].fillna(0, inplace=True)
df['ShotPower'].fillna(0, inplace=True)
df['Jumping'].fillna(0, inplace=True)
df['Stamina'].fillna(0, inplace=True)
df['Strength'].fillna(0, inplace=True)
df['LongShots'].fillna(0, inplace=True)
df['Aggression'].fillna(0, inplace=True)
df['Interceptions'].fillna(0, inplace=True)
df['Positioning'].fillna(0, inplace=True)
df['Vision'].fillna(0, inplace=True)
df['Penalties'].fillna(0, inplace=True)
df['Composure'].fillna(0, inplace=True)
df['Marking'].fillna(0, inplace=True)
df['StandingTackle'].fillna(0, inplace=True)
df['SlidingTackle'].fillna(0, inplace=True)
df['GKDiving'].fillna(0, inplace=True)
df['GKHandling'].fillna(0, inplace=True)
df['GKKicking'].fillna(0, inplace=True)
df['GKPositioning'].fillna(0, inplace=True)
df['GKReflexes'].fillna(0, inplace=True)

Luego de sustituír los valores vacíos, se verifica que el data frame esté libre de "missing values".

El primer sum() Sumariza por columna y el segundo vuelve a sumarizar y obtiene un dato escalar.

In [55]:
df.isnull().sum().sum()

0